src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs

```diff
@@ -43,21 +42,66 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
                 out this._fields);
         }
 
+        /// <summary>
+        /// Copy the original tuple, but modify it to use the new underlying type.
+        /// </summary>
         private TupleTypeSymbol(TupleTypeSymbol originalTuple,
                                 NamedTypeSymbol newUnderlyingType)
         {
             _underlyingType = newUnderlyingType;
-            _fields = originalTuple._fields.SelectAsArray(field =>
-                                                          {
-                                                              var newUnderlyingField = field.UnderlyingTypleFieldSymbol.OriginalDefinition.AsMember(_underlyingType);
-                                                              return new TupleFieldSymbol(field.Name, this, newUnderlyingField.Type, newUnderlyingField);
-                                                          });
+
+            ArrayBuilder<NamedTypeSymbol> underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            GetUnderlyingTypeChain(newUnderlyingType, underlyingTupleTypeChain);
```

Comment:
We could probably do without collecting all the types into the builder. The types are needed in the same order, as they are chained in type arguments, so we could just dig when appropriate as we iterate through the fields.